### ISSUE TITLE
Minimal example needed to call a Teams User from a Bot (application hosted media)

### ISSUE THREAD

**Issue by sipsorcery** [microsoftgraph/microsoft-graph-comms-samples#267]
**Describe the issue**
I have been unable to find a code sample that shows how a bot can initiate a call to a Teams user. The examples in this repo show how a bot can receive a call. An example showing how to initiate calls is also needed.

**Code Snippet**
My best attempt at the code to initiate a call is below. The sample does run and get a `Not Found ` response from the server. The configuration parameters I'm using do work correctly to allow a bot to receive a call from Teams.

````csharp
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Threading.Tasks;
using Microsoft.Graph.Communications.Common.Telemetry;
using Microsoft.Extensions.Logging;
using Microsoft.Graph;
using Microsoft.Graph.Communications.Calls;
using Microsoft.Graph.Communications.Calls.Media;
using Microsoft.Graph.Communications.Client;
using Microsoft.Skype.Bots.Media;

namespace caller
{
    class Program
    {
        private static string APP_NAME = "";
        private static string APP_ID = "";
        private static string APP_SECRET = "";
        private static string TENANT_ID = "";
        private static string CALLBACK_URI = "";
        private static string CERTIFICATE_THUMBPRINT = "";
        private static int MEDIA_PORT = 10000;
        private static string PUBLIC_IP = "";
        private static string HOSTNAME = "";

        static async Task Main(string[] args)
        {
            Console.WriteLine("Teams Call Console:");

            GraphLogger graphLogger = new GraphLogger(APP_NAME);
            graphLogger.DiagnosticLevel = System.Diagnostics.TraceLevel.Verbose;
            ILogger logger = new ConsoleLogger(graphLogger);

            AuthenticationProvider authProvider = new AuthenticationProvider(
                APP_NAME,
                APP_ID,
                APP_SECRET,
                TENANT_ID,
                graphLogger);

            GraphServiceClient graphClient = new GraphServiceClient(authProvider);

            var users = await graphClient.Users.Request().GetAsync();

            foreach (var user in users)
            {
                Console.WriteLine($"user Id: {user.Id}.");
                Console.WriteLine($"user Display Name: {user.DisplayName}.");
            }

            var mediaPlatformSettings = new MediaPlatformSettings()
            {
                MediaPlatformInstanceSettings = new MediaPlatformInstanceSettings()
                {
                    CertificateThumbprint = CERTIFICATE_THUMBPRINT,
                    InstanceInternalPort = MEDIA_PORT,
                    InstancePublicIPAddress = IPAddress.Parse(PUBLIC_IP),
                    InstancePublicPort = MEDIA_PORT,
                    ServiceFqdn = HOSTNAME,
                },

                ApplicationId = APP_ID,
            };

            var builder = new Microsoft.Graph.Communications.Client.CommunicationsClientBuilder(
                APP_NAME,
                APP_ID,
                graphLogger);

            builder
                .SetAuthenticationProvider(authProvider)
                .SetNotificationUrl(new Uri(CALLBACK_URI))
                .SetMediaPlatformSettings(mediaPlatformSettings)
                .SetServiceBaseUrl(new Uri(CALLBACK_URI));

            var client = builder.Build();

            Console.WriteLine($"Attempting to call {users.First().DisplayName}.");

            var call = new Call
            {
                CallbackUri = CALLBACK_URI,
                TenantId = TENANT_ID,
                Targets = new List<InvitationParticipantInfo>()
                {
                    new InvitationParticipantInfo
                    {
                        Identity = new IdentitySet
                        {
                            User = new Identity
                            {
                                DisplayName = users.First().DisplayName,
                                Id = users.First().Id
                            },
                        }
                    }
                },
                RequestedModalities = new List<Modality>()
                {
                    Modality.Audio
                },
                MediaConfig = new AppHostedMediaConfig(),
            };

            var callResult = await client.Calls().AddAsync(call);

            Console.WriteLine($"Call result {callResult.Id}.");

            Console.WriteLine("Finished.");
            Console.WriteLine("Press any key to exit...");
            Console.ReadLine();
        }

        /// <summary>
        /// Creates the local media session.
        /// </summary>
        /// <param name="mediaSessionId">
        /// The media session identifier.
        /// This should be a unique value for each call.
        /// </param>
        /// <returns>The <see cref="ILocalMediaSession"/>.</returns>
        private ILocalMediaSession CreateLocalMediaSession(ICommunicationsClient client)
        {
            // create media session object, this is needed to establish call connectionS
            var mediaSession = client.CreateMediaSession(
                new AudioSocketSettings
                {
                    StreamDirections = StreamDirection.Sendrecv,
                    SupportedAudioFormat = AudioFormat.Pcm16K
                },
                mediaSessionId: default(Guid));

            return mediaSession;
        }
    }
}
````

**Expected behavior**
I'm the only user on my Teams tenant so the expected behaviour is that my Teams client alerts me to an incoming call.

**Graph SDK (please complete the following information):**
 - Version 1.0 
   - Microsoft.Graph.Communications.Calls 1.2.0.850
   - Microsoft.Graph.Communications.Calls.Media 1.2.0.850
   - Microsoft.Graph.Communications.Client 1.2.0.850

**Logs**
````
Teams Call Console:
AuthenticationProvider: Info AuthenticationProvider: Generating OAuth token.
AuthenticationProvider: Info AuthenticationProvider: Generated OAuth token. Expires in 59.9825520633333 minutes.
user Id: 863a4f33-21a4-4586-9d5b-7839xxxxxxx.
user Display Name: xxx.
CommunicationsClient: Info Assembly: GraphCommunicationsClient
CommunicationsClient: Info Application: xxx
CommunicationsClient: Info Product: GraphCommunicationsClient-xxx
CommunicationsClient: Info Version: 1.2.0.850
CommunicationsClient: Info UserAgent: GraphCommunicationsClient-xxx/1.2.0.850
Attempting to call xxx.
CommunicationsClient: Info Dispatcher Container registering notification dispatcher for: communications/calls
CommunicationsClient: Verbose StatefulCallCollection creating subscription: communications/calls
NotificationDispatcher: Info Subscribed: communications/calls
To queue: communications/calls
CommunicationsClient: Info Media session was not provided, using media configuration defined in the call.
StatefulCall: Info Created new GraphLoggerShim for StatefulCall.
StatefulCall: Info Created new GraphAuthClient for StatefulCall.
StatefulCall: Info Authenticating request for tenant: 23c8477b-251f-4272-82c4-f2xxxxxx
AuthenticationProvider: Info AuthenticationProvider: Generating OAuth token.
AuthenticationProvider: Info AuthenticationProvider: Generated OAuth token. Expires in 59.847680015 minutes.
StatefulCall: Info {
  "@odata.type": "#microsoft.graph.call",
  "direction": "outgoing",
  "callbackUri": "https://xxx.com:5001",
  "source": {
    "@odata.type": "#microsoft.graph.participantInfo",
    "identity": {
      "@odata.type": "#microsoft.graph.identitySet",
      "application": {
        "@odata.type": "#microsoft.graph.identity",
        "displayName": "(Identity: Ihpd+/1p)",
        "id": "1e07c545-97f3-4f0b-b390-8e21xxxxxx"
      }
    }
  },
  "targets": [
    {
      "@odata.type": "#microsoft.graph.invitationParticipantInfo",
      "identity": {
        "@odata.type": "#microsoft.graph.identitySet",
        "user": {
          "@odata.type": "#microsoft.graph.identity",
          "displayName": "(Identity: F5S7njoJ)",
          "id": "863a4f33-21a4-4586-9d5b-78390xxxxxx"
        }
      }
    }
  ],
  "requestedModalities": [
    "audio"
  ],
  "mediaConfig": {
    "@odata.type": "#microsoft.graph.appHostedMediaConfig"
  },
  "tenantId": "23c8477b-251f-4272-82c4-fxxxxxxx"
}
StatefulCall: Verbose 
StatefulCall: Info 
StatefulCall: Verbose 
StatefulCall: Info 
StatefulCall: Info 
StatefulCall: Error {
  "error": {
    "code": "itemNotFound",
    "message": "Unexpected exception returned from the service.\r\nStatus Code: NotFound"
  }
}
StatefulCall: Info 
````

**Comment by sipsorcery** [comment]
Or does this note on the [Create call](https://docs.microsoft.com/en-us/graph/api/application-post-calls?view=graph-rest-1.0&tabs=http) reference page mean calls to a Teams user are not supported unless they have a VoIP number assigned?

> Note: For a call with app-hosted media, you need the Calls.AccessMedia.All permission in addition to one of the permissions listed in the table above.

**Comment by sipsorcery** [comment]
Bump.

Any chance of even a yes/no answer on this?

Is it possible to use the Graph SDK + Skype Bot SDK to initiate a call from a Bot to a Teams User?

**Comment by sipsorcery** [comment]
@ssulzer sorry for the ping but you seem to be the main person able to answer questions about the these samples and API's they use. I've got a project hanging on finding an answer to this question.

Is it supported to call a Teams user from a Bot or are there certain caveats about a VoIP number being assigned or such? The documentation is very vague, e.g. what does "VoIP number assigned mean"?

**Comment by ssulzer** [comment]
@ananmishr @zhengni-msft Can you comment please.

Here's a previous discussion about making outbound bot -> Teams user calls: https://github.com/microsoftgraph/microsoft-graph-comms-samples/issues/17#issuecomment-431183218

**Comment by sipsorcery** [comment]
Thanks to `ssulzer` the linked issue in the previous post I was able to get an additional pointer in the right direction and the `Not Found` error response I was getting was due to incorrect processing in my web callback. After a few more trial and error attempts I managed to get my bot to ring my Teams User (audio not wired up as of yet). That has provided answers to my two questions:

> Is it possible to use the Graph SDK + Skype Bot SDK to initiate a call from a Bot to a Teams User?

Yes.

> Is it supported to call a Teams user from a Bot or are there certain caveats about a VoIP number being assigned or such?

A VoIP number is NOT required to be assigned to the Teams User in order to receive a call.
